### PR TITLE
Build Vehicle Window

### PR DESCRIPTION
**Pull request by duncanspumpkin**
Complete. Will also add the draw track tab function.

### CODE REVIEW

File: src/openloco/windows/build_vehicle.cpp [hunk 1]
```diff
@@ -122,4 +463,519 @@ namespace openloco::ui::build_vehicle
...
+    {
+        if (widgetIndex < widx::tab_vehicles_for_0 || widgetIndex >= widx::scrollview_1)
+        {
+            *_common_format_args = string_ids::tooltip_scroll_new_vehicle_list;
+        }
+        else
+        {
+            auto unk_for_tab = widget_index_to_tab_vehicle_for(widgetIndex);
+            auto type = _11364F0[unk_for_tab];
+            if (type == 0xFF)
```
Comment (AaronVanGeffen) [inline src/openloco/windows/build_vehicle.cpp:701]: Linux and macOS CI are complaining here. `_11364F0` is defined as an `int8_t`, so 0xFF is out of range.
```suggestion
            if (type == -1)
```

File: src/openloco/windows/build_vehicle.cpp [hunk 2]
```diff
@@ -1,27 +1,274 @@
...
+
+    loco_global<uint16_t[8], 0x112C826> _common_format_args;
+    static loco_global<string_id, 0x009C68E8> gGameCommandErrorTitle;
+    static loco_global<int16_t, 0x01136268> _num_available_vehicles;
+    static loco_global<uint16_t[objectmgr::get_max_objects(object_type::vehicle)], 0x0113626A> _available_vehicles;
+    static loco_global<uint16_t, 0x0113642A> _113642A;
+    static loco_global<int32_t, 0x011364E8> _build_target_vehicle;
+    static loco_global<uint32_t, 0x011364EC> _11364EC;
+    // Array of types if 0xFF then no type, flag (1<<7) as well
+    static loco_global<int8_t[widget_index_to_tab_vehicle_for(widx::tab_vehicles_for_7) + 1], 0x011364F0> _11364F0;
```
Comment (duncanspumpkin) [inline src/openloco/windows/build_vehicle.cpp:95]: track type for tab

File: src/openloco/windows/build_vehicle.cpp [hunk 3]
```diff
@@ -1,27 +1,274 @@
...
+            unk_al |= (1 << 7);
+            if (unk_al == 0xFF)
+            {
+                unk_al = _525FC5;
+            }
+        }
+
+        bool found = false;
+        uint32_t unk_tab_offset = 0;
+        for (; unk_tab_offset < _11364EC;)
```
Comment (duncanspumpkin) [inline src/openloco/windows/build_vehicle.cpp:255]: This is finding the tab with a compatible track type for this vehicle

File: src/openloco/windows/build_vehicle.cpp [hunk 4]
```diff
@@ -122,4 +480,839 @@ namespace openloco::ui::build_vehicle
...
+        }
+
+        window->var_874 = found ? ebp : 0;
+
+        bool unk_flag = _11364F0[ebp] & (1 << 7);
+        uint8_t cl = _11364F0[ebp] & ~(1 << 7);
+        sub_4A3A06(cl, unk_flag);
+    }
+
+    static void sub_4A3A06(uint8_t cl, bool unk_flag)
```
Comment (duncanspumpkin) [inline src/openloco/windows/build_vehicle.cpp:1208]: unk_flag `isRoad`

File: src/openloco/windows/build_vehicle.cpp [hunk 5]
```diff
@@ -122,4 +480,839 @@ namespace openloco::ui::build_vehicle
...
+                    args.push(string_ids::airport);
+                }
+                else
+                {
+                    args.push(string_ids::docks);
+                }
+            }
+            else
+            {
+                bool unk_flag = type & (1 << 7);
```
Comment (duncanspumpkin) [inline src/openloco/windows/build_vehicle.cpp:732]: `isRoad`

File: src/openloco/windows/build_vehicle.cpp [hunk 6]
```diff
@@ -1,27 +1,283 @@
...
+        caption = 1,
+        close_button = 2,
+        panel = 3,
+        tab_build_new_trains,
+        tab_build_new_buses,
+        tab_build_new_trucks,
+        tab_build_new_trams,
+        tab_build_new_aircraft,
+        tab_build_new_ships,
+        tab_vehicles_for_0,
```
Comment (duncanspumpkin) [inline src/openloco/windows/build_vehicle.cpp:40]: `track_tab`. Go through all of this and conform to track_tab and track type for this set of tabs.

File: src/openloco/objects/interface_skin_object.h [hunk 7]
```diff
@@ -161,17 +161,17 @@ namespace openloco
 
         constexpr uint32_t toolbar_menu_industries = 191;
 
-        constexpr uint32_t build_vehicle_train = 251;
+        constexpr uint32_t build_vehicle_train = 251; // 15 frames
```
Comment (AaronVanGeffen) [inline src/openloco/objects/interface_skin_object.h:164]: Fwiw, we've so far been using image ids for all frames involved. See e.g. `ScenarioOptions::drawTabs`.

**Comment by duncanspumpkin** [comment]
Bug with changing primary tab whilst on a non zero secondary tab.

File: src/openloco/windows/build_vehicle.cpp [hunk 8]
```diff
@@ -120,11 +120,11 @@ namespace openloco::ui::build_vehicle
...
-    static void sub_4C28D2(ui::window* window);
-    static void sub_4C2D8A(ui::window* window);
-    static void sub_4C1CBE(ui::window* window);
-    static void sub_4A3A06(uint8_t track_type, bool is_road);
-    static void sub_4C2865(ui::window* window);
+    static void SetDisabledTransportTabs(ui::window* window);
+    static void SetTrackTypeTabs(ui::window* window);
+    static void ResetTrackTypeTabSelection(ui::window* window);
+    static void SetTopToolbarLastTrack(uint8_t track_type, bool is_road);
+    static void SetTransportTypeTabs(ui::window* window);
```
Comment (AaronVanGeffen) [inline src/openloco/windows/build_vehicle.cpp:127]: I really appreciate you naming these functions. However, I would like to continue to use camelCase for function names to distinguish them from types (which should use TitleCase).

File: src/openloco/windows/build_vehicle.cpp [hunk 9]
```diff
@@ -1,27 +1,285 @@
...
+    static void get_scroll_size(ui::window* window, uint32_t scrollIndex, uint16_t* scrollWidth, uint16_t* scrollHeight);
+    static void on_scroll_mouse_down(ui::window* window, int16_t x, int16_t y, uint8_t scroll_index);
+    static void on_scroll_mouse_over(ui::window* window, int16_t x, int16_t y, uint8_t scroll_index);
+    static void tooltip(FormatArguments& args, ui::window* window, widget_index widgetIndex);
+    static ui::cursor_id cursor(window* window, int16_t widgetIdx, int16_t xPos, int16_t yPos, ui::cursor_id fallback);
+    static void prepare_draw(ui::window* window);
+    static void draw(ui::window* window, gfx::drawpixelinfo_t* dpi);
+    static void draw_scroll(ui::window* window, gfx::drawpixelinfo_t* dpi, uint32_t scrollIndex);
+
+    static void initEvents()
```
Comment (AaronVanGeffen) [inline src/openloco/windows/build_vehicle.cpp:149]: I generally put this function at the bottom of the namespace, so the event functions do not have to be predeclared. It's not something that needs changing, per se, but something to keep in mind, perhaps.

File: src/openloco/windows/build_vehicle.cpp [hunk 10]
```diff
@@ -86,40 +455,954 @@ namespace openloco::ui::build_vehicle
...
+        auto companyColour = companymgr::get_company_colour(window->number);
+
+        for (auto tab : _transportTypeTabInformation)
+        {
+            auto frameNo = 0;
+            if (_transportTypeTabInformation[window->current_tab].type == tab.type)
+            {
+                frameNo = (window->frame_no / 2) & 0xF;
+            }
+            uint32_t image = gfx::recolour(skin->img + tab.imageId + frameNo, companyColour);
```
Comment (AaronVanGeffen) [inline src/openloco/windows/build_vehicle.cpp:1304]: While this is likely the most efficient manner, for the other windows we defined a lookup table for all animation frames. The compiler is able to optimise this, and it leaves room for the images to not be in sequence.

I don't feel too strongly about this, but @marijnvdwerf asked me to change mine to lookup tables in the past. I'll leave it to you.

Comment (duncanspumpkin) [reply src/openloco/windows/build_vehicle.cpp:1304]: Yes i see that now looking through the other works. Will change it.

**Pull request merged**